roxctl/analyze/README.md

```diff
@@ -0,0 +1,95 @@
+It produces a list of **allowed** connections based on the workloads and network policies defined.
+It is based on [NP-Guard's Network Policy Analyzer](https://github.com/np-guard/netpol-analyzer) component.
+For more details, refer to the [NP-Guard webpage](https://np-guard.github.io/).
+
+## Command Objective
+Generate a file that allows users to visualize the connectivity posture induced by the network policies generated by `roxctl generate netpol`.
+
+## Usage
+
+### Analyzing allowed connectivity from YAML manifests (network policies and workload resources) 
+To analyze network policies, `roxctl analyze netpol` requires a folder containing Kubernetes manifests, including network policies.
+The manifests must not be templated (e.g., Helm charts) to be considered.
+All YAML files that could be accepted by `kubectl apply -f` will be accepted as a valid input and searched by `roxctl analyze netpol`.
+
+Example run with output to `stdout`:
+
+```shell
+$ roxctl analyze netpol  tests/roxctl/bats-tests/test-data/np-guard/netpols-analysis-example-minimal/
+0.0.0.0-255.255.255.255 => default/frontend[Deployment] : TCP 8080
+default/backend[Deployment] => default/backend[Deployment] : All Connections
+default/frontend[Deployment] => 0.0.0.0-255.255.255.255 : UDP 53
+default/frontend[Deployment] => default/backend[Deployment] : TCP 9090
+default/frontend[Deployment] => default/frontend[Deployment] : All Connections
+```
+
+Example output with `md` output format:
+```shell
+$ roxctl analyze netpol  tests/roxctl/bats-tests/test-data/np-guard/netpols-analysis-example-minimal/ -o md
+```
+
+| src | dst | conn |
+|-----|-----|------|
+| 0.0.0.0-255.255.255.255 | default/frontend[Deployment] | TCP 8080 |
+| default/backend[Deployment] | default/backend[Deployment] | All Connections |
+| default/frontend[Deployment] | 0.0.0.0-255.255.255.255 | UDP 53 |
+| default/frontend[Deployment] | default/backend[Deployment] | TCP 9090 |
+| default/frontend[Deployment] | default/frontend[Deployment] | All Connections |
+
+
+#### Understanding the output
+
+The output contains a list of permitted connectivity lines. Each connectivity line is of the format `src => dst : connnectivity-attributes`.
+The `src, dst` can be any of the analyzed cluster workloads or an IP address range.
+If both `src` and `dst` are cluster workloads, it means that both `src` is allowed to send traffic to `dst` and that `dst` is allowed to receive traffic from `src`, over the specified connectivity attributes.
+If one of `src` or `dst` are IP address ranges, it means that only one direction of this connection (either egress from workload or ingress to workload) is explicitly permitted by network policies.
+ 
+For example, in the above output, the line: `default/frontend[Deployment] => 0.0.0.0-255.255.255.255 : UDP 53` specifies permitted egress from `frontend` to any IP address.

```

Comment:
I think it might be a bit misleading, since the cluster may have workloads not included in the analyzed dir and not protected by network policies (e.g. `kube-system` pods) , for which their IP address can overlap with an IP range in the output. 

This sentence only refers to one direction of  **permitted egress traffic** , for which it is true to say that any IP address is allowed. 
To explain about the connection in both directions (egress from `frontend` and ingress to any IP address from the range), it depends on the type of the destination:
- destination outside the cluster is not in the scope of the analysis (thus we assume the connection is allowed)
- destination in the cluster but not included in the workloads of the analyzed dir and not protected by network policies  (e.g. `kube-system` pods)  - the connection is allowed if such workload's IP address overlaps with the IP range in the output.
- destination in the cluster and included in the analyzed dir: the connection is allowed only if there is a specific line in the output with a connection from `frontend` to that workload. 

I think if we want to specify when the connection between `frontend`to IP range is allowed and not only the egress part, 
a more accurate (but maybe confusing) clarification is "to any IP address outside of the cluster or to in-cluster workloads which are not analyzed and not protected by network policies".

Reply:
I asked this and suggested an explanation, because I think, there may be a chance of misunderstanding when reading this. Look at the following two sentences:

1.  _specifies permitted egress from `frontend` to any IP address_
2. _It does not mean that a connection from `frontend` to `backend` is allowed_

The issue here is the contradiction of `any` from 1. and a concrete exclusion in 2. (all but `backend`). If `any` IP address should exclude the IP of the `backed`, then I think, we should try to rephrase this a bit to make it more clear.

I would propose something myself, but honestly I am also wrapping my head around this and I struggle to explain it clearly.

Reply:
The 'any' IP address should not exclude the IP of the `backend`.  
There is a sentence explaining the "one direction constraint" (only ingress or egress can be inferred as allowed) in a connection to/from IP ranges:  (as opposed to a connection between two workloads): 

_If one of `src` or `dst` are IP address ranges, it means that only one direction of this connection (either egress from workload or ingress to workload) is explicitly permitted by network policies._

Thus, sentence number (1) that you mentioned does not imply that any IP address is going to accept incoming traffic from `frontend`. It only implies that `frontend` is allowed to send egress traffic to any IP address.

Sentence number (2) clarifies this constraint on this example. The entire connection (considering both ingress and egress aspects) from `frontend` to `backend` is not allowed, because of the ingress constraint on `backend` (although the egress from `frontend` is allowed).

The goal was to clarify the different semantics of a connection in the output between two workloads (that considers both ingress and egress constraints) , and a connection between a workload to ip-range (that considers only ingress/egress direction).

Reply:
Okay, I understand it after your explanation, thank you :) The communication between `frontend` and `backend` would fail, because the `backend` would block the traffic, despite `frontend` allowing it.

I am okay with leaving the text as it is, but if you would like to capture some parts of our discussion into text, then please go ahead. Otherwise I would merge this soon.

Reply:
Thanks @vikin91 , I think we can leave it as it is now.